File: js/bin/canon_check.js
// Reads the shared corpus (spec/corpus/canon.json) and confirms this JS
// canonicalizer reproduces each vector's expected bytes. The Java, Elixir, Rust,
// and Go implementations validate the same file, so agreement means a
// certificate signed by any of them verifies under this one. Invoked by
// interop/check-canon-js.sh; exits non-zero on any mismatch.
import fs from 'node:fs';
import { canonicalize } from '../src/canon.js';

const path = process.argv[2];
if (!path) { process.stderr.write('usage: canon_check <canon.json>\n'); process.exit(2); }
const doc = JSON.parse(fs.readFileSync(path, 'utf8'));

let failures = 0;
for (const v of doc.vectors) {
  let got;
  try {
    got = canonicalize(v.cert);
  } catch (e) {
    console.log(`FAIL ${v.name}\n  error: ${e.message}`);
    failures++;
    continue;
  }
  if (got === v.canonical) {
    console.log(`PASS ${v.name}`);
  } else {
    console.log(`FAIL ${v.name}\n  got:  ${got}\n  want: ${v.canonical}`);
    failures++;
  }
}
if (failures > 0) { process.stderr.write(`${failures} vector(s) mismatched\n`); process.exit(1); }
console.log(`all ${doc.vectors.length} canon vectors match`);

File: rust/src/bin/canon_check.rs
//! Cross-language interop check: reads the shared corpus
//! (spec/corpus/canon.json) and confirms this Rust canonicalizer reproduces
//! each vector's expected bytes. The Java, Go, and Elixir implementations
//! validate the same file, so agreement means a certificate signed by any of
//! them verifies under this one. Invoked by interop/check-canon-rust.sh; exits
//! non-zero on any mismatch.

use std::process::exit;

use bonemesh::canon::canonicalize;
use serde_json::Value;

fn main() {
    let path = std::env::args().nth(1).unwrap_or_else(|| {
        eprintln!("usage: canon_check <path-to-canon.json>");
        exit(2);
    });

    let text = std::fs::read_to_string(&path).expect("read corpus");
    let doc: Value = serde_json::from_str(&text).expect("parse corpus");
    let vectors = doc["vectors"].as_array().expect("vectors array");

    let mut failures = 0;
    for v in vectors {
        let name = v["name"].as_str().unwrap_or("?");
        let got = canonicalize(&v["cert"]).expect("canonicalize");
        let want = v["canonical"].as_str().expect("canonical");
        if got == want {
            println!("PASS {name}");
        } else {
            println!("FAIL {name}\n  got:  {got}\n  want: {want}");
            failures += 1;
        }
    }

    if failures > 0 {
        eprintln!("{failures} vector(s) mismatched");
        exit(1);
    }
    println!("all {} canon vectors match", vectors.len());
}
